lib/index.js

```diff
@@ -403,36 +459,66 @@ exports.getFunctionUrl = async (ctx, boundaryId, functionId) => {
 
 exports.getStorageClient = (ctx) => {
     const expiry = 60 * 16; // 15+1 min to align with Lambda lifecycle plus some buffer
-    const accessToken = Jwt.sign({}, Buffer.from(ctx.configuration.fusebit_storage_key, 'base64').toString('utf8'), {
-        algorithm: 'RS256',
-        expiresIn: expiry,
-        audience: ctx.configuration.fusebit_storage_audience,
-        issuer: ctx.configuration.fusebit_storage_issuer_id,
-        subject: ctx.configuration.fusebit_storage_subject,
-        keyid: ctx.configuration.fusebit_storage_key_id,
-        header: { jwtId: Date.now().toString() },
-    });
+    let accessToken = ctx.fusebit && ctx.fusebit.functionAccessToken;
+    if (ctx.configuration.fusebit_storage_key) {
+        // Legacy mode - use pre-creaeted issuer to mint the access token
+        accessToken = Jwt.sign({}, Buffer.from(ctx.configuration.fusebit_storage_key, 'base64').toString('utf8'), {
+            algorithm: 'RS256',
+            expiresIn: expiry,
+            audience: ctx.configuration.fusebit_storage_audience,
+            issuer: ctx.configuration.fusebit_storage_issuer_id,
+            subject: ctx.configuration.fusebit_storage_subject,
+            keyid: ctx.configuration.fusebit_storage_key_id,
+            header: { jwtId: Date.now().toString() },
+        });
+    }
 
     const url = generateStorageUrl(ctx.configuration);
 
-    return {
+    const isHierarchicalStorageSupported = accessToken === ctx.fusebit.functionAccessToken;
+
+    const ensureHierarchicalStorageSupported = () => {
+        if (!isHierarchicalStorageSupported) {
+            throw new Error('Hierarchical storage is not supported in this function.');
+        }
+    }
+
+    const getUrl = (storageSubId) => `${url}${storageSubId ? ('/' + storageSubId) : ''}`;
+
+    const storageClient = {
         etag: null,
```

Comment:
@tjanczuk The stateful `etag` property on this class is making less and less sense with our guidance to basically never cache the storage client, which is causing me to effectively treat it as a method object that I throw away every use. I am basically having to cache the etag externally after every get and put, and supply it via this property before every put. I wonder if we should just be upfront, return it in the get payload, and ask users to explicitly supply it in the put payload and eliminate this property.

Reply:
Not storing etags in client any more - returning with data and accepting on put.